File: 101/04 二分查找/02_34.cc
class Solution {
  public:
    vector<int> searchRange(vector<int>& nums, int target) {
      if (nums.empty()) return vector<int>{-1, -1};
      int lower = LowerBound(nums, target, true);
      int upper = LowerBound(nums, target, false) - 1;
      if (lower == nums.size() || nums[lower] != target) return vector<int>{-1, -1};
      return vector<int>{lower, upper};
    }

    int LowerBound(vector<int>& nums, int target, bool lower) {
      int l = 0;
      int r = nums.size();
      int mid = -1;
      while (l < r) {
        mid = l + (r - l) / 2;
        bool condition = lower ? (nums[mid] >= target) : (nums[mid] > target);
        if (condition) {
          r = mid;
        } else {
          l = mid + 1;
        }
      }
      return l;
    }
};

File: 101/04 二分查找/02_34.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
  if (nums.length === 0) return [-1, -1];
  const lower = lowerBound(nums, target, true);
  const upper = lowerBound(nums, target, false) - 1;
  if (lower === nums.length || nums[lower] !== target) return [-1, -1];
  return [lower, upper];
};

function lowerBound(nums, target, lower) {
  let l = 0;
  let r = nums.length;
  let mid = -1;
  while (l < r) {
    mid = l + Math.floor((r - l) / 2);
    condition = lower ? (nums[mid] >= target) : (nums[mid] > target);
    if (condition) {
      r = mid;
    } else {
      l = mid + 1;
    }
  }
  return l;
}
